### PR TITLE
Indexed solution storage

### PR DESCRIPTION
**Pull request by IngridKJ**
## Proposed changes

In preparation for higher-order time-stepping, storage of multiple solutions (both time-step and iterate solutions) is now supported. 

Data storage structure for indexed solution storage is included. In practice this means pp.STATE and pp.ITERATE are removed and replaced with pp.TIME_STEP_SOLUTIONS and pp.ITERATE_SOLUTIONS, respectively. Time step solutions and iterate solutions are stored in separate sub-dictionaries within data-dictionary. This means that iterate solutions are no longer stored inside the sub-dictionary for time step solutions (previously pp.STATE). Structure of data is thus:
data[time_step_solution or iterate_solution][variable name][storage index]

As a consequence of the removal of pp.STATE/pp.ITERATE, also the methods (and belonging tests) set_state and set_iterate is removed. They are replaced by a utility function set_solution_values (found in porepy/numerics/ad/equation_system.py), which is now used heavily in various tests.

For handling the multiple stored solutions, new methods for shifting already stored values within the data dictionary is included. These new methods and documentation of them are found within the EquationSystem class.

## Types of changes

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] Testing (contribution related to testing of existing or new functionality).
- [x] Documentation (contribution related to adding, improving, or fixing documentation).
- [x] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

- [x] The documentation is up-to-date.
- [x] Static typing is included in the update.
- [x] This PR does not duplicate existing functionality.
- [x] The update is covered by the test suite (including tests added in the PR).

### CODE REVIEW

**Review by IvarStefansson — COMMENTED** [review]
Very good! Please have a look at my suggestions and questions, fix and report back.

**Comment by IvarStefansson** [comment]
We had several conflicts in exporter.py. @jwboth, could you please have a look at that particular file to confirm that we haven't messed anything up?

File: tests/unit/test_rt0.py [hunk 1]
```diff
@@ -1230,7 +1230,7 @@ def test_rt0_2d_simplex_surf(self):
 
     def test_convergence_rt0_2d_iso_simplex(self):
 
-        a = 8 * np.pi ** 2
+        a = 8 * np.pi**2
```
Comment (keileg) [inline tests/unit/test_rt0.py:1233]: I guess this and severeal other random changes in the tests is caused by running black. As a rule, I would prefer if we do so in dedicated PRs since this will make the review much simpler: We are not allowed to argue with black so that PR is accepted by default, while for standard PRs, the reviewer will not have to filter away random formatting edits.

**Review by keileg — APPROVED** [review]
Nice work!

Minor comments only. I am fine with merging once they have been addressed.

File: src/porepy/viz/exporter.py [hunk 2]
```diff
@@ -750,12 +751,11 @@ def write_pvd(
         pvd_file: Path = Path(
             self._append_folder_name(self._folder_name, self._file_name) + ".pvd"
         )
-        file_exists: bool = pvd_file.exists()
+        zfile_exists: bool = pvd_file.exists()
```
[the remark this answers is not in the document]
Comment (IngridKJ) [reply src/porepy/viz/exporter.py:754]: This seems to have happened by an accident, so from my point of view it should be ok that I revert the change
